Add supports for feature importances

yufongpeng opened the pull request:
I've added three methods for calculating feature importances.
The default `feature_importances` was calculating by `Mean Decrease in Impurity` which is calculated simultaneously with model building.
`permutation_importances` shuffles the columns multiple times and compares `R2` or `accuracy` with the original model. 
`dropcol_importances` deletes each columns and uses cross validation instead.

ablaom commented:
@yufongpeng Thanks for this contribution. A cursory review has convinced me that this is worthy and well-executed enhancement to DecisionTrees.jl. 

I think I can make a good detailed review of this PR but it will a little time. To help with any review, can you please add document strings for the three new feature importance methods, and also a document string for your enhancement of `prune_tree` (which you did not mention above)?

Additionally, we will need to add tests to this PR. The current tests do not really test the added implementations, as far as I can tell.

yufongpeng commented:
There's some changes for this branch:
1. I move `dropcol_importances` to `scikeatLearnAPI`. Cross validation is optional, but I can not test cv unless there is a function to do cv with built-in `nfoldCV` on sklearn wrappers.
2. `permutation_importances` can only do multiple shuffles for native trees but can do cv for sklearn wrappers. However, as same as `dropcol_importances`, this can not be tested.
3. The modification for `prune_tree` is just recalculating `feature_importances` (modifying property `featim`) if necessary.
4. Docstrings and more tests are added. 
I use `similarity` (cosine of two importance vectors) to quantify their similarity; I'm not sure if this is adequate.
I tend to compare same kind of feature importances between models with different hyperparameters, but some are not available, especially for randomforests and adabooststumps. Some of these tests are very robust but many are quite unstable between julia versions, so I deprecate a lot of such tests.

ablaom commented on src/classification/tree.jl:
```diff
@@ -21,15 +21,18 @@ module treeclassifier
         region      :: UnitRange{Int}   # a slice of the samples used to decide the split of the node
         features    :: Vector{Int}      # a list of features not known to be constant
         split_at    :: Int              # index of samples
+        ni          :: Float64
```
I suppose `ni` is node impurity. Can we rename this `node_impurity` or `impurity`. Or, at the very least, add a comment here saying what it is.

ablaom commented on src/DecisionTree.jl:
```diff
@@ -40,19 +42,44 @@ struct Node{S, T}
+    featval :: S
+    left    :: Union{Leaf{T}, Node{S, T}}
+    right   :: Union{Leaf{T}, Node{S, T}}
+    featim  :: Vector{Float64}
```
```suggestion
    featim  :: Vector{Float64}   # MDI feature importance
```

ablaom commented on Project.toml:
```diff
@@ -12,6 +12,7 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 ScikitLearnBase = "6e75b9c4-186b-50bd-896f-2d2496a4843e"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
```
What are we needing from StatsBase?

ablaom commented on src/classification/main.jl:
```diff
@@ -48,14 +48,28 @@ function _convert(
+    end
+    return 
+end
+
```
Can we call this `get_node_impurity!`  or `get_impurity!`?

ablaom replied:
Or maybe better, `update_using_impurity!`

ablaom commented on src/classification/main.jl:
```diff
@@ -48,14 +48,28 @@ function _convert(
         weights      = nothing;
-        rng          = Random.GLOBAL_RNG) where {S, T}
+        rng          = Random.GLOBAL_RNG,
+        calc_fi             :: Bool = true) where {S, T}
```
Can we please have a more user-friendly name to replace `calc_fi`. This is pretty challenging for non-English speakers. Why not `feature_importance`? I don't see an issue with conflict with function of same name, as this is a kwarg. 

Or, if you don't like that, how about `calculate_importances` or `calculate_feature_importances`.

ablaom commented on src/measures.jl:
```diff
@@ -307,3 +316,84 @@ function nfoldCV_forest(
+    normalize ? trees.featim ./ sum(trees.featim) : trees.featim
+
+feature_importances(trees::T; kwargs...) where {T <: Ensemble} = isdefined(trees, :featim) ? trees.featim : Float64[]
+feature_importances(tree::T; kwargs...) where {T <: DecisionTree.Node} = Float64[]
```
Perhaps I missed something, but it appears you do not expose in the public API a `feature_importance` for a *single* tree; it always returns empty. Is there a reason for that?

yufongpeng replied:
If a single tree was built with `calc_fi` setting  true (I will change to `feature_importances`), the type would be `RootNode`, and `featim` was defined. If  `calc_fi` was false,  the type would be `Node` without `featim`.

ablaom commented on src/classification/main.jl:
```diff
@@ -101,32 +126,72 @@ function build_tree(
+        get_ni!(fi, t.root)
+        return RootNode{S, T}(t.root.feature, t.root.threshold, left, right, fi ./ size(features, 1))
+    end
 end
```
This code looks identical to that for `build_stump`. Just a suggestion, but maybe you want to factor that out to reduce code duplication.

ablaom commented on src/classification/main.jl:
```diff
@@ -101,32 +126,72 @@ function build_tree(
 end
 
-function prune_tree(tree::LeafOrNode{S, T}, purity_thresh=1.0) where {S, T}
+function prune_tree(tree::LeafOrNode{S, T}, purity_thresh=1.0, loss = util.entropy) where {S, T}
```
I still think it would be good to document what how this `prune_tree` is using the feature importance.

ablaom replied:
Also, there's a lot of logic here which is not tested for correctness in the test suite.

ablaom commented on src/classification/main.jl:
```diff
@@ -298,13 +383,14 @@ function build_adaboost_stumps(
     coeffs = Float64[]
     for i in 1:n_iterations
-        new_stump = build_stump(labels, features, weights; rng=rng)
+        new_stump = build_stump(labels, features, weights; rng=mk_rng(rng), calc_fi=false)
```
For AdaBoost you are defining the feature importance differently, using the coefficients, right? So that should be appear in one of the doc-strings, right?

yufongpeng replied:
I have a problem in the algorithm for adaboost:
The coeff is 1/2 log((1+err)/(1-err)), which is minimized at err = 0. I think this should be 1/2 log((1-err)/err) + log(n_label -1 (to prevent 0) which is negatively correlated with err.

ablaom replied:
Are you saying there is a bug? If so, great catch! but please open separate PR to resolve and I will prioritise a review.

ablaom commented on src/measures.jl:
```diff
@@ -307,3 +316,84 @@ function nfoldCV_forest(
+
+Return an vector of feature importances calculated by `Mean Decrease in Impurity (MDI)`.
+If `calc_fi` was set false, this function returns an empty vector.
+
```
I have found some confusion about what exactly MDI means. [This post](https://medium.com/the-artificial-impostor/feature-importance-measures-for-tree-models-part-i-47f187c1a2c3) (my top hit for googling "MDI feature importance") suggests that sometimes the impurity decreases are just summed over nodes, but sometimes they are weighted by number of samples, and so forth. I therefore suggest we spell out explicitly what we are doing. Here's a suggestion:

Feature importances are computed as follows:

- Single tree: For each feature the associated importance is the sum, over all splits based on that feature, of the impurity decreases for that split (the node impurity less the sum of the child impurities) divided by the total number of training observations.
- Forests: Assuming `normalise==false`, the importance for a given feature is the average over trees in the forest of the tree importances for that feature. If `normalize=true`, then ...
- AdaBoost models: The features importance is ...

ablaom commented on src/regression/main.jl:
```diff
@@ -42,8 +52,17 @@ function build_tree(
+        get_ni!(fi, t.root)
+        return RootNode{S, T}(t.root.feature, t.root.threshold, left, right, fi ./ size(features, 1))
+    end
 end
```
Here's that repeating code block again (I think) from classification/main.jl

yufongpeng replied:
This part is a little different from classification/main.jl;  it does not have  the second `list` argument.

ablaom replied:
Ah, yes, thanks for the clarification.

ablaom commented on src/regression/main.jl:
```diff
@@ -99,11 +120,29 @@ function build_forest(
+        end
+
+        return Ensemble{S, T}(forest_new, fi ./ n_trees)
+    end
```
I think this block is identical to one added for forests in classification/main.jl

ablaom reviewed (commented):
@yufongpeng Thanks for this extra work.

I now believe that this repository is not the right place to be implementing model-generic feature importances, like `dropcol_importances` and `permutation_importances`. I'm sorry for not flagging this earlier. It would be more useful to include this code where it could be applied to any number of supervised learning models (@OkonSamuel is close to adding the scaffolding for feature importance to [MLJ.jl](https://alan-turing-institute.github.io/MLJ.jl/dev/). Perhaps you would you consider contributing your code there instead? I expect we will add a new package, MLJFeatureSelection or something like that.) At the very least, the model-generic importances should be factored out into a separate PR.

I have reviewed the tree-specific feature importance code and have not reviewed the model-generic feature importance code. My comments are mostly about readability and some suggestions to reduce code duplication, and a request for more detailed documentation. Historically this package has dumped all the docs into the readme, but moving forward, I would like to see more in-line documentation.

Thanks for adding the testing. Unfortunatlely, these are only consistency tests (and require the extra model-generic feature importances) not thorough checks for correctness. I think the most importance thing is to check the importances for a single tree are correct. I suggest we do one (ideally both) of the following:

- explicitly define  a small data set which will lead to a decision tree classifier / regressor of depth 3 or 4 that one can derive by hand, with all randomness switched off in our choice of hyper-parameters. Check the implementation predicts importances computed by hand for this tree. 
- compare the decision tree feature importances on some larger data set for the current implementation with an identical implementation elsewhere, which is already well-tested, eg, scikitlearn (again with randomness off) and check agreement. 

I'm open to other suggestions for correctness tests.

yufongpeng commented:
I've added both split and impurity based methods for feature importance and deleted part of permutation importance and dropcol imortance which is model-generic. Now, permutation importance only permutate each features and can not do cross validation.
There is a new test file for testing feature importance. An artificial data containing 20 samples (10 0s and 10 1s) and 3 features is used to test the exact value of both split importance and impurity importance including pruned tree. For permutation importance and ensemble methods, the data was enlarged by a factor of 50 with a little randomness. 
`digits` and `iris` are used to compare with `scikit-learn`. I'm only able to get identical models in both `DecisionTree.jl` and `scikit-learn` for the first two tree method, so this two models have much more strict tests.
Currently, tests for adaboost models are not on because those importances are not correct under current algorithm. After merging #167, I'll enable them.

ablaom commented on test/classification/scikitlearn.jl:
```diff
@@ -23,9 +23,10 @@ model = fit!(RandomForestClassifier(), features, labels)
-@test impurity_importance(model) == impurity_importance(model.ensemble, model.coeffs)
-@test split_importance(model) == split_importance(model.ensemble, model.coeffs)
-@test isapprox(permutation_importance(model, features, labels, rng = 1).mean, permutation_importance((model.ensemble, model.coeffs), labels, features, (model, y, X) -> accuracy(y, apply_adaboost_stumps(model, X)), rng = 1).mean)
+# On when switching to SAMME
```
Note to self: Need re-instating tests below after #167

ablaom reviewed (approved):
@yufongpeng Thanks for the very substantial improvements to this already great PR. The new testing is especially appreciated. I can see this would have been a considerable investment of your time. 🙏🏾

I have no further requests of substance. Waiting for:

- [ ] merge #167 
- [ ] re-instate AdaBoost tests as per comment above

rikhuijzer commented on src/classification/main.jl:
```diff
@@ -163,14 +230,16 @@ function apply_tree(tree::LeafOrNode{S, T}, features::AbstractMatrix{S}) where {
 
 `col_labels` is a vector containing the distinct labels
 (eg. ["versicolor", "virginica", "setosa"]). It specifies the column ordering
 of the output matrix. """
```
```suggestion
of the output matrix.
"""
```

rikhuijzer commented on src/classification/main.jl:
```diff
@@ -163,14 +230,16 @@ function apply_tree(tree::LeafOrNode{S, T}, features::AbstractMatrix{S}) where {
 of the output matrix. """
+apply_tree_proba(tree::Root{S, T}, features::AbstractVector{S}, labels) where {S, T} = 
+    apply_tree_proba(tree.node, features, labels)
 apply_tree_proba(leaf::Leaf{T}, features::AbstractVector{S}, labels) where {S, T} =
```
```suggestion
apply_tree_proba(leaf::Leaf, features::AbstractVector, labels) =
```

ablaom commented:
@yufongpeng `dev` branch now contains the SAMME update. Ready for you to rebase and re-instate those tests that relied on that update. Please also address @rikhuijzer's comments.

yufongpeng commented:
@ablaom I think this PR is almost ready, but there's one thing I need to fix. In the new version of the `prune_tree`, I only consider classification model and update `impurity_importance` based on entropy (or other discrete metric like gini) because `prune_tree` is based on accuracy. In the test file, there's  a test on regression tree using `prune_tree` and the result of `impurity_importance` is definitely wrong. I'm able to fix this but just not sure if using accuracy on regression tree is OK and we should avoid using prune on a regression tree.

ablaom commented on src/classification/main.jl:
```diff
@@ -102,31 +129,71 @@ function build_tree(
+
+Prune tree based on prediction accuracy of each nodes.
+* `purity_thresh`: If prediction accuracy of a stump is larger than this value, the node will be pruned and become a leaf.
+* `loss`: The loss function for computing node impurity. It can be either `util.entropy`, `util.gini` or other measures of impurity depending on the loss function used for building the tree. 
```
I'm finding this one sentence a little confusing. Do you mean to say that the loss can be any of these, but would generally be chosen to match the loss used in building the tree (but doesn't have to be)?

yufongpeng replied:
Yes, you can change the loss function freely because the tree did not store the loss function and it has to be provided externally. I've changed the algorithm a little bit: If a tree `Root{S, T}` where T is `Float64` , the loss function will be `mean_squared_error`; otherwise, it will be `util.entropy`.

ablaom commented on src/classification/main.jl:
```diff
@@ -102,31 +129,71 @@ function build_tree(
+Prune tree based on prediction accuracy of each nodes.
+* `purity_thresh`: If prediction accuracy of a stump is larger than this value, the node will be pruned and become a leaf.
+* `loss`: The loss function for computing node impurity. It can be either `util.entropy`, `util.gini` or other measures of impurity depending on the loss function used for building the tree. 
+Impurity importances will be recalculated by substracting impurity decrease of the pruned node divided by total number of training observations when the tree has `featim` property.
```
You mention "impurity importances" here, but do not say how they are used in the pruning algorithm, as far as I can  tell. 

Also, is the use of purity importances  (for some or any choice of `loss`) a breaking change in the original algorithm? This seems to be the case, but I want to confirm this for the release notes.

yufongpeng replied:
I've improved the doc. 

Yes, I think it is a breaking change because if the original tree was built using `util.gini`, this specific function must be provided, or the `featim` will be wrongly updated.

ablaom replied:
The improvement is clearer. But I'd still like to see an explicit explanation of how purity importances are used to prune the tree.

yufongpeng replied:
I think there's some misunderstanding. Pruning is still based on accuracy and purity does not affect how the tree is pruned. It's the opposite that impurity importance will be affected by pruning, so the algorithm I added is just updating the new impurity importance without changing any behavior of `prune_tree` itself.

ablaom replied:
Oh, now I see! Yes that makes sense now. Thanks for the clarification.

ablaom commented:
@yufongpeng I agree that an accuracy-based  `prune_tree` does not make sense for regression and we can remove that test. Please add a comment to the `prune_tree` doc-string that warns against use in regression.

ablaom commented on src/classification/main.jl:
```diff
@@ -144,16 +165,19 @@ function _build_tree(tree::treeclassifier.Tree{S, T}, labels::AbstractVector{T},
 Prune tree based on prediction accuracy of each nodes.
 * `purity_thresh`: If prediction accuracy of a stump is larger than this value, the node will be pruned and become a leaf.
-* `loss`: The loss function for computing node impurity. It can be either `util.entropy`, `util.gini` or other measures of impurity depending on the loss function used for building the tree. 
+* `loss`: The loss function for computing node impurity. For classification tree, default function is `util.entropy`; for regression tree, it's `mean_squared_error`. 
```
Two points: 

- `util.entropy` is not in the `DecisionTree` namespace. So unless we decide to export this, we should write `DecisionTree.util.entropy`

- We should mention `DecisionTree.util.gini` as an available option, as user is unlikely to discover it otherwise.

kescobo commented:
Apologies if this is orthogonal to the PR itself, and also if this is a naive question, but I'm anxious to get some results that make use of it. Are changes required to MLJDecisionTreeInterface in order for me to use this via MLJ? I've added @yufongpeng 's fork, and trained a machine using the MLJ interface. I tried using `impurity_importance()` on the machine, as well as `mach.model`, but neither worked.

Is there some other way to do this, or do I need to use the DecisionTree API directly instead?

OkonSamuel commented:
> Apologies if this is orthogonal to the PR itself, and also if this is a naive question, but I'm anxious to get some results that make use of it. Are changes required to MLJDecisionTreeInterface in order for me to use this via MLJ? I've added @yufongpeng 's fork, and trained a machine using the MLJ interface. I tried using `impurity_importance()` on the machine, as well as `mach.model`, but neither worked.
> 
> Is there some other way to do this, or do I need to use the DecisionTree API directly instead?

I'm currently working on MLJ API for feature importance. I'll be done with it this week.

kescobo commented:
Got it, thanks for putting this together. I was able to use it on the `DecisionTree.jl` model directly.

rikhuijzer reviewed (changes requested):
The diff doesn't only show what this PR changes but also contains the changes of #174 and other PRs. This is probably caused by a wrong merge of `dev` into this PR. I don't know how to fix it via Git, but you could copy the files over to a new PR based on `dev` and then it should all be looking good again

OkonSamuel commented on src/measures.jl:
```diff
@@ -307,3 +317,149 @@ function nfoldCV_forest(
+    base = score(trees, labels, features)
+    scores = Matrix{Float64}(undef, size(features, 2), n_iter)
+    rng = mk_rng(rng)::Random.AbstractRNG
+    for (i, col) in enumerate(eachcol(features))
```
`eachcol` methods requires at least julia 1.1. I suggest we bump the minimum julia compat for DecisionTrees to "1.1".
@ablaom @yufongpeng.

rikhuijzer replied:
It is already at 1.6 (#179) but this PR is outdated.

kescobo commented:
>I don't know how to fix it via Git, but you could copy the files that you've changed over to a new PR based on dev and then it should all be looking good again

I just attempted a rebase, but there was a lot going on that I didn't totally understand. Someone is going to have to be careful to not undo stuff that's happened on `dev` since this PR was started.

ablaom commented:
@yufongpeng Can you please try @rikhuijzer 's [suggestions](https://github.com/JuliaAI/DecisionTree.jl/pull/166#pullrequestreview-1022995774) to clean up the git issues?

ablaom commented:
Closed in favour of #182
